br/pkg/storage/azblob.go

```diff
@@ -324,6 +390,82 @@ type AzureBlobStorage struct {
 
 	cpkScope *blob.CPKScopeInfo
 	cpkInfo  *blob.CPKInfo
+
+	// resolvedAccountName is the final account name we are going to use.
+	resolvedAccountName     string
+	resolvedServiceEndpoint string
+}
+
+// CopyFrom implements Copier.
+func (s *AzureBlobStorage) CopyFrom(ctx context.Context, e ExternalStorage, spec CopySpec) error {
+	es, ok := e.(*AzureBlobStorage)
+	if !ok {
+		return errors.Annotatef(berrors.ErrStorageInvalidConfig,
+			"AzureBlobStorage.CopyFrom supports *AzureBlobStorage only, got %T", e)
+	}
+
+	url, err := urlOfObjectByEndpoint(es.resolvedServiceEndpoint, es.options.Bucket, es.withPrefix(spec.From))
+	if err != nil {
+		return errors.Annotatef(err, "failed to get url of object %s", spec.From)
+	}
+	dstBlob := s.containerClient.NewBlobClient(s.withPrefix(spec.To))
+
+	// NOTE: `CopyFromURL` supports files up to 256 MiB, which might not be enough for huger regions.
+	// Hence we use the asynchronous version and wait for finish.
+	// But this might not as effect as the syncrhonous version for small files.
+	// It is possible to use syncrhonous version for small files if necessary.
+	// REF: https://learn.microsoft.com/en-us/rest/api/storageservices/copy-blob-from-url
+	resp, err := dstBlob.StartCopyFromURL(ctx, url, &blob.StartCopyFromURLOptions{})
+	if err != nil {
+		return errors.Annotatef(err, "failed to copy blob from %s to %s", url, spec.To)
+	}
+	copyID := resp.CopyID
+	deref := aws.StringValue
+
+	for {
+		prop, err := dstBlob.GetProperties(ctx, &blob.GetPropertiesOptions{})
+		if err != nil {
+			return errors.Annotate(err, "failed to check asynchronous copy status")
+		}
+		if prop.CopyID == nil || deref(prop.CopyID) != deref(copyID) {
+			return errors.Annotatef(berrors.ErrStorageUnknown,
+				"failed to check copy status: copy ID not match; copy id = %v; resp.copyID = %v;", deref(copyID), deref(prop.CopyID))
+		}
+		cStat := deref((*string)(prop.CopyStatus))
+		// REF: https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob-properties?tabs=microsoft-entra-id#response-headers
+		switch cStat {
+		case "success":
+			return nil
+		case "failed", "aborted":
+			return errors.Annotatef(berrors.ErrStorageUnknown, "asynchronous copy failed or aborted: %s", deref(prop.CopyStatusDescription))
+		case "pending":
+			finished, total, err := progress(deref(prop.CopyProgress))
+			if err != nil {
+				return errors.Annotate(err, "failed to parse progress")
+			}
+			rem := total - finished
+			toSleep := time.Duration(rem/azblobPremisedCopySpeedPerMilliSecond) * time.Millisecond
```

Comment:
Should we need to consider the concurrency? If there are 5 blob-copy requests to the server, the speed for each copy progress may be slower.

Reply:
It seems polling more frequently isn't pretty harmful. But you just notified me a minimal backoff time might be needed or tons of requests will bring many bills...